GH-6170: Search bar: the chosen word from the drop-down is not shown

benmarcotte opened the pull request:
Closes #6170

benmarcotte committed:
feat: searchbox matches highlighted result

benmarcotte committed:
feat: searchbox displays selected result

benmarcotte committed:
feat: search bar shows selected result

benmarcotte committed:
feat: search box reflects selected result

benmarcotte committed:
Merge branch 'develop' into feature/GH-6170

benmarcotte committed:
feat: seach box reflects selected item

benmarcotte committed:
Merge branch 'feature/GH-6170' of https://github.com/SAP/spartacus into feature/GH-6170

qjimmy committed:
swap let keyword for const

bgambocjaviniar reviewed (changes requested):
Don't think we need the pipe nor searching through the focused element when clicking to find if it exist in the array.

What you can do is look into ViewChild and get the selected suggestion value from there, in which you can use that value to populate the input with

----
The pipe is throwing the error right away when the application is just open in console
The pipe throws error when clicking a value as well

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
feat: reflect chosen word in search bar

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

saad-mo commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -46,6 +46,7 @@ export class SearchBoxComponent {
    * for example when we click inside the search result section.
    */
   private ignoreCloseEvent = false;
+  public chosenWord = '';
```
This should be private, per our coding guidelines - https://sap.github.io/spartacus-docs/coding-guidelines/#services

saad-mo replied:
since this member needs to be accessible in the template, it cannot be private.

bgambocjaviniar replied:
let's omit the public keyword, however do note that if it needed to be private, you can always use getters/setters

saad-mo reviewed (changes requested):
This PR addresses the issue at hand, but also introduces a small UI bug as well. When you navigate away from the search results page, the key is still retained in the search box. The search box should be cleared when you navigate away from the search results page. 

Update: After reviewing the behavior on dev branch, this is a pre-existing UI bug and it's best to address in this PR.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
GH-6170 remove searched word when nevigated away

saad-mo commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -56,8 +58,17 @@ export class SearchBoxComponent {
+    protected winRef: WindowRef,
+    protected router: Router
+  ) {
+    this.router.events.subscribe((event: Event) => {
```
We need to `unsubscribe` from this observable in `ngOnDestroy()`

saad-mo reviewed (changes requested):
- We will need unit test to make sure the chosen word in the search box is same as in the URL
- Also another unit test to check if the search box is empty when not in the search results view

Parthlakhani committed:
reafactor: unsuscribe the subscription

Parthlakhani committed:
GH-6170 add unit tests

Parthlakhani committed:
refactor: remove focus from describe

Parthlakhani committed:
refactor: add check fot input in unit tests

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

saad-mo reviewed (dismissed):
Looks good, thanks!

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -56,8 +62,20 @@ export class SearchBoxComponent {
-    protected winRef: WindowRef
-  ) {}
+    protected winRef: WindowRef,
+    protected routingService: RoutingService,
```
breaking change

will have a call with you soon

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -56,8 +62,20 @@ export class SearchBoxComponent {
+      .getRouterState()
+      .subscribe((data) => {
+        const search = this.routingConfigService
+          .getRouteConfig('search')
```
we can do something better than the logic applied here (will discuss together in a call in relation to the breaking change)

Basically, look at the param query and context of the page through the routingService instead to reset the chosen word

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -46,7 +51,8 @@ export class SearchBoxComponent {
   private ignoreCloseEvent = false;
-
+  public chosenWord = '';
+  public updateChosenWordFromUrl: Subscription;
```
let's keep the variable naming consistent as it is throughout the application

change to `subscription`

bgambocjaviniar replied:
just putting the link to the above comment https://github.com/SAP/spartacus/pull/8298#discussion_r568719587 as you might miss it

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -232,4 +254,8 @@ export class SearchBoxComponent {
     });
   }
+
+  ngOnDestroy() {
```
I don't know how this is working, but it needs to implement OnDestroy

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts:
```diff
@@ -7,12 +7,15 @@ import {
 import { SearchBoxComponentService } from './search-box-component.service';
 import { SearchBoxComponent } from './search-box.component';
 import { SearchResults } from './search-box.model';
+import { RouterState } from './../../../../../core/src/routing/store/routing-state';
```
this should be included in the @spartacus/core import

Parthlakhani committed:
feat: Change logic to remove chosen word (GH-6170)

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -46,17 +52,35 @@ export class SearchBoxComponent {
+    routingService: RoutingService
+  );
+  /**
+   * @deprecated
```
As mentioned in the DM, it would be best to add comments, such as the constructor deprecation in `projects/storefrontlib/src/cms-components/navigation/search-box/search-box-component.service.ts`

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -89,6 +113,15 @@ export class SearchBoxComponent {
+    this.subscription = this.routingService
+      .getRouterState()
+      .subscribe((data) => {
+        if (!data.nextState && data.state.params?.query !== this.chosenWord)
```
1) Any reason to keep the `!data.nextState`? What's the use case for it?

2) As mentioned before, wouldn't it be better to include `data.state.context.type === PageType.CONTENT_PAGE && data.state.context.id === 'search'` in order to be sure that you're really in the search page. 
As I'm not sure if xxx.params?.query is the only use case for the search page. This property may exist in other places. (not sure, but it's a possibility)

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -89,6 +113,15 @@ export class SearchBoxComponent {
     switchMap((config) => this.searchBoxComponentService.getResults(config))
   );
 
+  ngOnInit() {
```
let's explicitly add `:void`

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -159,6 +192,10 @@ export class SearchBoxComponent {
     return <HTMLElement>this.winRef.document.activeElement;
   }
 
+  updateChosenWord(chosenWord: string) {
```
let's explicitly add `:void`

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts:
```diff
@@ -232,4 +269,8 @@ export class SearchBoxComponent {
     });
   }
+
+  ngOnDestroy() {
```
let's explicitly add `:void`

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts:
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
+        navigationId: null,
+      };
+      const input = fixture.debugElement.query(By.css('input'));
+      const PRODUCT_SEARCH_STRING = 'camera';
```
let's move this as 'global' as it's being used in 2 test cases

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts:
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
     });
 
+    it('should contain chosen word from the dropdown', () => {
+      const mockRouterState = {
```
let's move this object as 'global' and setting the params property as an empty object to be used here https://github.com/SAP/spartacus/pull/8298/files#diff-8713f6e9795265679e836cf3c74e49434601908590027faf3be387cb6e4ee5d0R72

After, you can just re-define an object using that mockedObject and spread it with the value you have in this current params

bgambocjaviniar commented on projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts:
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
+      fixture.detectChanges();
+      expect(searchBoxComponent.chosenWord).toEqual(PRODUCT_SEARCH_STRING);
+      expect(
+        fixture.debugElement.query(By.css('input')).nativeElement.value
```
`fixture.....(by.css('input'))` can be replaced by `input`

bgambocjaviniar replied:
wasn't changed, but let's get on a call to address some of the new changes done

Parthlakhani committed:
Merge branch 'develop' of https://github.com/SAP/spartacus into feature/GH-6170

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
refactor: Update logic for updateing chosen word (GH-6170)

Changed logic for updating chosen word when going away from search page.
Updated unit tests accoordingly and moved local variables to global.

Parthlakhani committed:
Merge branch 'develop' into feature/GH-6170

Parthlakhani committed:
refactor: Update condition for updating chosen word
